framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -306,10 +306,25 @@
     private final Map<Tree, AnnotatedTypeMirror> classAndMethodTreeCache;
 
     /**
-     * Mapping from a Tree to its annotated type; before implicits are applied, just what the
-     * programmer wrote.
+     * Mapping from an expression tree to its annotated type; before implicits are applied, just
```

Comment:
Why all the changes to caching?

Reply:
This issue is only triggered with a small cache. The reason why this worked with a large cache is that getting types for Expressions, Members, and Types used the same cache. An earlier request for a type tree was then found when an expression query happened. With a small cache that earlier value gets evicted.
To prevent this kind of problem completely, this changes it into three separate caches. We should never again run into a situation where something gets cached as an expression and then returned for a TypeTree query.
That would seem like a big win.

Reply:
Makes sense.  Thanks for the explanation!